DatadogRUM/Sources/Instrumentation/Actions/SwiftUI/SwiftUIComponentDetector.swift

```diff
@@ -0,0 +1,146 @@
+    func createActionCommand(
+        from touch: UITouch,
+        predicate: SwiftUIRUMActionsPredicate?,
+        dateProvider: DateProvider
+    ) -> RUMAddUserActionCommand?
+}
+
+internal enum SwiftUIComponentFactory {
+    /// Factory that creates the appropriate SwiftUI detector based on platform and version.
+    /// Modern detection is only available on iOS 18+ and tvOS 18+.
+    static func createDetector() -> SwiftUIComponentDetector {
+        if #available(iOS 18.0, tvOS 18.0, visionOS 18.0, *) {
+            return ModernSwiftUIComponentDetector()
+        }
+        return LegacySwiftUIComponentDetector()
+    }
+}
+
+/// Utility class with static helper methods
+internal class SwiftUIComponentHelpers {
+    /// Extracts a component name from a touch, using gesture information when available
+    /// - Parameters:
+    ///   - touch: The `UITouch` to analyze
+    ///   - defaultName: The fallback name to use if a more specific one can't be determined
+    /// - Returns: The best available component name
+    ///
+    /// Note: SwiftUI component detection relies on internal implementation details
+    /// which may change with iOS updates. This method uses various heuristics to
+    /// identify common components like buttons and navigation links.
+    /// Note: We check gestures' names both during the `.began` and `.ended` phases
+    /// as we can collect different information.
+    static func extractComponentName(touch: UITouch, defaultName: String) -> String {
+        // Check gesture recognizers' names for more specific info
+        if let gestures = touch.gestureRecognizers {
+            for gesture in gestures {
+                if let gestureName = gesture.name {
+                    // E.g., Button<ResolvedButtonStyleBody<BorderlessButtonStyleBase>>
+                    if gestureName.hasPrefix("Button<") {
+                        return SwiftUIComponentNames.button
+                    }
+                }
+            }
+        }
+
+        return defaultName
+    }
+
+    /// Extracts a SwiftUI Toggle action from a touch if applicable
+    /// - Parameters:
+    ///   - touch: The `UITouch` to analyze
+    ///   - predicate: The predicate to use for determining if an action should be created
+    ///   - dateProvider: Provider for current time
+    /// - Returns: A RUM action command if a toggle was detected, `nil` otherwise
+    static func extractSwiftUIToggleAction(
+        from touch: UITouch,
+        predicate: SwiftUIRUMActionsPredicate?,
+        dateProvider: DateProvider
+    ) -> RUMAddUserActionCommand? {
+        guard let predicate = predicate,
+              let view = touch.view else {
+            return nil
+        }
+
+        #if !os(tvOS)
+        let uiSwitch = view.findInParentHierarchy { parent in
+            // Note: we might want to extend to `UIControl` in the future.
+            return parent is UISwitch
+        }
+
+        if uiSwitch != nil,
+           touch.phase == .ended,
```

Comment:
This `touch.phase == .ended` should be the first statement of the guard above, to not loop through parent if it's ignored anyway.

Reply:
That being said, I wonder why we check for the `touch.phase` in here. This method is only about extracting the name, the phase is already checked in the caller.